### ISSUE TITLE
using JSON-optimized-data-strucutere versus the existing YANG XML-optimized-data-strucutres in REST

### ISSUE THREAD

**Issue by slieberth** [openconfig/reference#20]
Hi Anees

Standards are essential to the success of networking, and unless there is major improvement by deviating from the standard, it is best practice to follow standards.

Have there been thoughts to use a JSON-optimized-data-strucutere instead of the XML-optimized-YANG data-strucutres. I think Netconf and Yang and bound closely together. Netconf requires XML. Now, If we move to REST APIs, which offers JSON support, a more efficient data-structure could be applied:

In the examples below:
JSON optimized code has 10 lines and 6 levels of hierarchy.
the YANG compliant equivalent has 18 lines and 9 levels of hierarchy.
That is an imporvemnt of 30-40%, also the JSON model is much easier to read.

e.g. this JSON data-strucutre (in YAML represenation)

``` json
bgp:
  neighbors:
    33.33.33.33:
      description: BGP-neighbor-description-3
      peer-as: 3333
      peer-type: external
      afi-safis:
        ipv4-unicast:
          enabled: true
          max-prefixes: 900001
          restart-timer: 61
```

has the same information content as the YANG-OpenConfig compliant datastructure (also in YAML format):

``` json
bgp:
  neighbors:
  - neighbor:
      neighbor-address: 33.33.33.33
      config:
        description: BGP-neighbor-description-3
        peer-as: 3333
        peer-type: external
      afi-safis:
      - afi-safi:
          afi-safi-name: ipv4-unicast
          config:
            enabled: true
          ipv4-unicast:
            prefix-limit:
              config:
                max-prefixes: 900001
                restart-timer: 61
```

Basically the JSON model would not use YANG lists, but just containers, leaf and leaf-Lists.

feedback welcome ... 

best regards Stefan

**Comment by robshakir** [comment]
Thanks for this suggestion -- the historical decision that was made here was to be compliant with the JSON serialisation that was described by the IETF for YANG-described data structures in RFC7951.

If there were to be more work on re-visiting this encoding, I think looking at the benefits over and above 7951, and whether JSON serialisation would be the right form (vs. some binary encoding) would be merited.

Closing this issue -- please feel free to re-open it if you have further discussion points here.
